lib/Sema/CSSimplify.cpp

```diff
@@ -1410,6 +1410,22 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
     case ConstraintKind::ArgumentConversion:
     case ConstraintKind::OperatorArgumentTupleConversion:
     case ConstraintKind::OperatorArgumentConversion:
+      if (type1->is<InOutType>() && typeVar2 &&
+          typeVar2->getImpl().canBeInOut()) {
+        // Upgrade type2 to an InOutType
+        auto inOutType1 = type1->getAs<InOutType>();
+        
+        unsigned typeVar3Options = typeVar2->getImpl().getOptions();
+        typeVar3Options &= ~TVO_CanBeInOut;
+        auto typeVar3 = createTypeVariable(getConstraintLocator(locator),
+                                           typeVar3Options);
+        auto inOutType3 = InOutType::get(typeVar3);
+        
+        addConstraint(ConstraintKind::Equal, typeVar2, inOutType3, locator);
+        return matchTypes(inOutType1->getObjectType(), typeVar3, kind, flags,
```

Comment:
This seems unnecessarily complex, with the introduction of a new type variable that is immediately bound. The effect of this code is, basically, binding `typeVar2` to `InOutType::get(inOutType1->getObjectType())`, i.e., `type1`. Why can't we directly allow that type binding once we've checked that the type variable can bind to an `InOutType`?